browser/views.py

```diff
@@ -687,3 +686,47 @@ def delete_investigator_record(request, person_id, research_id):
         Investigator.objects.filter(id=research_id).delete()
 
     return HttpResponseRedirect(reverse('person', args=(person_id,)))
+
+
+@staff_member_required
+def position(request, person_id, position_id=None):
+    person = get_object_or_404(Person, pk=person_id)
+    template = "browser/position.html"
+    form = PositionForm()
+    context = {
+        'form': form,
+        'person': person,
+        'position_id': None
+    }
+
+    if position_id:
+        position = get_object_or_404(Position, pk=position_id)
+        form = PositionForm(initial={'subject': position.subject, 'role': position.role,
+                                     'year': position.year, 'start_date': position.start_date,
+                                     'end_date': position.end_date})
+        context.update({
+            'position_id': position_id,
+            'form': form
+        })
+
+    if request.method == 'POST':
+        form = PositionForm(request.POST)
+        if form.is_valid():
+            if position_id is None:
+                Position.objects.create(
+                    person_id=person_id,
+                    changed_by=request.user,
+                    **form.cleaned_data
+                )
+
+            else:
+                Position.objects.filter(id=position_id).update(**form.cleaned_data)
+            return HttpResponseRedirect(reverse('person', args=(person.id,)))
+    return render(request, template, context)
```

Comment:
if the form is not valid, we still need it in the context, don't we? to be able to show errors.

Reply:
The form is in the context when we are initially calling the position function. So, the errors in the form are displayed to the user.

Reply:
but if there is a position id (so position is edited), then the form would be overwritten by line 704 and 709, doesn't it?

Reply:
When we are making a POST request , a call is made to fetch the form at line 713 - <br>``` form = PositionForm(request.POST) ```. <br>This gets the value which the user wants to update.
The lines 704 and 709 do set the form to the initial value. But , the above code retrieves the correct values from the POST request that needs to be updated.

Reply:
You are not returning a validated form (with errors) in the context. As a result, the template won't be able to display _why_ the POST request didn't go through.

Reply:
besides what Jitesh said, this is also unnecessarily very confusing. Just have a GET and a POST block, initialize the forms in them as needed (either with object, without, or from the POST request) and then update the context accordingly (in each block). There is no reason to initialize the form in line 695 if there is just one case where an empty form is needed.

Reply:
I have changed the implementation of the form. There is a GET and POST block, and the forms are being initialized within these blocks.